onnxruntime/python/tools/automl/data_frame_tool.py

```diff
@@ -0,0 +1,137 @@
+                    #print('Col: {} processed exact match type: {} '.format(input_meta.name, df[input_meta.name].dtype))
+                    feeds[input_meta.name] = np.array([df[input_meta.name][0]]).astype(expected_type).reshape(shape)
+                elif expected_type in ort_float_set and str(df[input_meta.name].dtype) in pd_float_set:
+                    #print('Col: {} processed as floating type: {} '.format(input_meta.name, df[input_meta.name].dtype))
+                    feeds[input_meta.name] = np.array([df[input_meta.name][0]]).astype(expected_type).reshape(shape)
+                elif expected_type in ort_int_set and str(df[input_meta.name].dtype) in pd_int_set:
+                    #print('Col: {} processed as integer type: {} '.format(input_meta.name, df[input_meta.name].dtype))
+                    feeds[input_meta.name] = np.array([df[input_meta.name][0]]).astype(expected_type).reshape(shape)
+                else:
+                    raise TypeError("Input {} expected to be of type: {} got {} ".format(
+                                input_meta.name, expected_type, df[input_meta.name].dtype))
+            elif require:
+                raise RuntimeError("This model requires input {} of type {} but it is not found in the DataFrame".format(
+                               input_meta.name, types_dict[input_meta.type]))
+        return feeds
+
+
+    def _get_input_feeds(self, df, sess):
+        "Return a dictionary of input_name : a typed and shaped np.array of values"
+        """
+        This function accepts Pandas DataFrame as the first argument and onnxruntime
+        session with a loaded model. The function interrogates the model for the inputs
+        and matches the model input names to the DataFrame instance column names.
+        It requires exact matches for bool and float16 types. It attempts to convert to
+        string any input type if string is required.
+        It attempts to convert floating types to each other and does the same for all of the
+        integer types without requiring an exact match.
+
+        :param df: See :class:`pandas.DataFrame`. The function only considers the first row (0) of each column
+            and feeds the data to the appropriate model inputs.
+
+        :param sess: See :class:`onnxruntime.InferenceSession`.
+        
+        ::
+        For example: pd.DataFrame([[0], [4],[20]],index=[0], columns=['A', 'B', 'C'])
+
+        """
+        if df.empty:
+            raise RuntimeError('input DataFrame is empty')
+
+        # Process mandadory inputs. Raise an error if anything is not present
+        feeds = self._process_input_list(df, sess.get_inputs(), True)
+        initializers = self._process_input_list(df, sess.get_overridable_initializers(), False)
+
+        feeds.update(initializers)
+
+        return feeds
+
+    def execute(self, df, output_names=[], run_options=None):
```

Comment:
Due to wonky python rules, the default value for output_names will be used for every invocation of the method when a parameter value isn't provided; if the func modifies the list, it will only be empty the first time the method is called. For example:

def Func(my_list=[]):
    my_list.append("A")
    print(my_list)

Func() # Prints ["A"]
Func() # Prints ["A", "A"]

Because of this, a better way to write the default is:

def Func(my_list=None):
    my_list = my_list or []
    ...

This is one of Python's biggest warts as a language.